CI: format isort output for detection by azure pipelines

leeyspaul opened the pull request:
- [ ] closes #27179 
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Commit:
format isort output

datapythonista commented on ci/code_checks.sh:
```diff
@@ -109,7 +109,7 @@ if [[ -z "$CHECK" || "$CHECK" == "lint" ]]; then
     # Imports - Check formatting using isort see setup.cfg for settings
     MSG='Check import format using isort ' ; echo $MSG
-    isort --recursive --check-only pandas asv_bench
+    isort --recursive --check-only pandas asv_bench | awk -F ":" '##vso[task.logissue type=error;sourcepath=%(path)s]%(text)s'
```
Looks good, but we just want to change the format when the environment variable `AZURE` is set. You can see how this is done for flake8. Sorry it wasn't clear in the issue.

Can you also change the order of an import to see this failing in the CI please? So we can see it working before we merge. Thanks!

leeyspaul replied:
yes, thank you for being so helpful!

WillAyd replied:
Does this still grab the return code from `isort`?

Commit:
refactor isort formatting

Commit:
intentional misordering of imports to test ci failure

Commit:
change isort formatting

fix isort formatting for awk

fix isort format output

test awk change

fix isort output for azure

remove comment

fix deliberate import errors

fix whitespace

whitespace fix

leeyspaul commented:
@datapythonista sorry I'm a bit new to bash scripting. It's definitely catching the import error but it is ignoring the exception as shown here: 

```Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>
BrokenPipeError: [Errno 32] Broken pipe
Check import format using isort DONE
##[error]Bash exited with code '120'.```

leeyspaul commented:
pinging for review! Spent some time on it to understand `awk`. @datapythonista  thank you!

datapythonista commented on ci/code_checks.sh:
```diff
@@ -109,9 +109,12 @@ if [[ -z "$CHECK" || "$CHECK" == "lint" ]]; then
+    if [[ "$AZURE" == "true" ]]; then
+        isort --recursive --check-only pandas asv_bench | awk -F ':' '{print $0"\n" "##vso[task.logissue type=error;sourcepath=%(path)s]%(text)s"}'
+    else
+        isort --recursive --check-only pandas asv_bench | awk -F ':' '{print}'
```
I don't think we need `awk` in this case, do we?

datapythonista commented on ci/code_checks.sh:
```diff
@@ -109,9 +109,12 @@ if [[ -z "$CHECK" || "$CHECK" == "lint" ]]; then
     MSG='Check import format using isort ' ; echo $MSG
-    isort --recursive --check-only pandas asv_bench
+    if [[ "$AZURE" == "true" ]]; then
+        isort --recursive --check-only pandas asv_bench | awk -F ':' '{print $0"\n" "##vso[task.logissue type=error;sourcepath=%(path)s]%(text)s"}'
```
I think the return code `$?` we are checking later to see if `isort` was successful or not, will actually be the one of awk with this, which will always be 0.

This means that if the imports are not correctly sorted, we won't detect them in the CI.

Can you start by swapping the order of two imports in pandas in this PR, so we know if this is working or not? We'll remove them once the PR is approved.

leeyspaul replied:
Yes, I will push up the badly sorted imports

leeyspaul replied:
here is the output! ```Check import format using isort
ERROR: /home/vsts/work/1/s/pandas/plotting/_matplotlib/core.py Imports are incorrectly sorted.
##[error]%(path)s(,): error : %(text)s
Check import format using isort DONE
##[error]Bash exited with code '1'.
##[section]Finishing: Linting```

datapythonista commented on ci/code_checks.sh:
```diff
@@ -109,9 +109,12 @@ if [[ -z "$CHECK" || "$CHECK" == "lint" ]]; then
+        isort --recursive --check-only pandas asv_bench | awk -F ':' '{print}'
+    fi
     RET=$(($RET + $?)) ; echo $MSG "DONE"
-
```
in general try to avoid unrelated changes, even if they are small like this.

leeyspaul replied:
Undid the change!

datapythonista reviewed (changes requested):
Thanks @leeyspaul, getting closer, still some comments.

Commit:
address pr comments and deliberate bad imports ordering

fix output

test azuredev output

fix awk output to show path

reformat output to incorporate error msg

format output to catch path and error name

datapythonista commented:
Seems like `awk` is keeping the exit status of isort, I guess it's because of the `set -o pipefail` in the function at the top of the function, we'll have to fix that.

But the first things would probably be fixing the output, from:
```
ERROR: /home/vsts/work/1/s/pandas/plotting/_matplotlib/core.py 
##[error]%(path)s(,): error : %(text)s
```
to
```
##vso[task.logissue type=error;sourcepath=/home/vsts/work/1/s/pandas/plotting/_matplotlib/core.py] Imports are incorrectly sorted.
```

Thanks!

Commit:
fix error output substr

Commit:
formatting output to only have error msg

leeyspaul commented:
ok! I think I'm getting closer now. @datapythonista could you check out the failure? I've pasted below what it's returning for ease.

```
Check import format using isort
##[error]pandas/plotting/_matplotlib/core.py(,): error :   Imports are incorrectly sorted.Check import format using isort DONE
##[error]Bash exited with code '1'.
##[section]Finishing: Linting
```

I added the ORS as the error message was moving to a new line with print. If there is a better way to do this do let me know, thank you!

datapythonista commented:
is there a reason to not use the `##vso` code I suggested? I think is what we use in the other cases, if there is no reason better to keep consistent.

leeyspaul commented:
> is there a reason to not use the `##vso` code I suggested? I think is what we use in the other cases, if there is no reason better to keep consistent.

er, I didn't mean to remove anything. I'm not too familiar on vso codes but I read up on logging here: https://github.com/microsoft/azure-pipelines-tasks/blob/master/docs/authoring/commands.md

Could you clarify what you mean by the vso code and I will change it back, or point me to the docs. Thank you!

datapythonista commented:
I refer to what is suggested in https://github.com/pandas-dev/pandas/pull/27334#issuecomment-515995494 instead of using `##[error]`

leeyspaul commented:
ah right. So oddly it seems to be getting converted from `##vso` to `##[error]` though the awk line I've used is `isort --recursive --check-only pandas asv_bench | awk '{ORS=""} {print "##vso[task.logissue type=error;sourcepath="$2";]"} {$1=$2=""; print $0}'`.

I might be missing something obvious here

datapythonista commented:
ah, sorry, didn't see in the code you were already using the `##vso`. I guess that's some azure magic then, but will be applied to all the other commands too, so it's ok.

I'd prefer to remove the `awk` and leave the original call when not using azure. Will be misleading if the errors are different when people call `isort` directly than with our script I think.

And the thing that concerns me is the `set -o pipefail` inside the if at the beginning. If eventually we change that code, your code will fail. And that doesn't seem right, since both things are unrelated, and shouldn't be using side-effects of each other. I'd start by adding `set +o pipefail` after the grep in line 39. That should already be there, will avoid the side effects, and will make your code fail. After that, you can wrap your code with the same `set -o pipefail` and `set +o pipefail`, and that should fix it again.

Does this make sense?

leeyspaul commented:
> ah, sorry, didn't see in the code you were already using the `##vso`. I guess that's some azure magic then, but will be applied to all the other commands too, so it's ok.
> 
> I'd prefer to remove the `awk` and leave the original call when not using azure. Will be misleading if the errors are different when people call `isort` directly than with our script I think.
> 
> And the thing that concerns me is the `set -o pipefail` inside the if at the beginning. If eventually we change that code, your code will fail. And that doesn't seem right, since both things are unrelated, and shouldn't be using side-effects of each other. I'd start by adding `set +o pipefail` after the grep in line 39. That should already be there, will avoid the side effects, and will make your code fail. After that, you can wrap your code with the same `set -o pipefail` and `set +o pipefail`, and that should fix it again.
> 
> Does this make sense?

I think so, I'll take a look at this tonight! Thanks for the info. : )

WillAyd commented:
@leeyspaul can you merge master again? Old builds are no longer available

WillAyd commented:
Great! No worries on timing - not in any rush. Just wanted to make sure this was something that you were still looking at, so thanks for confirming

jreback commented:
@leeyspaul can you merge master and update

jbrockmendel commented:
@leeyspaul can you rebase

alimcmaster1 commented:
If we dont here from @leeyspaul I can take this on

leeyspaul commented:
@alimcmaster1 feel free to take over. Thanks!

alimcmaster1 commented:
Restarted this over here: https://github.com/pandas-dev/pandas/pull/29654

Closing this for now

datapythonista commented:
I'm moving the Checks build to actions, and I think it only supports the `##[error]` formatting, not the one with file, line number...

alimcmaster1 commented:
Ahh so there probably isn’t much value in this at all? I can close my PR if so

datapythonista commented:
There is surely value on letting the CI know what's an error. Just the format will probably be different.
